airflow-core/src/airflow/jobs/asyncio_monitor.py

```diff
@@ -0,0 +1,244 @@
+    return timezone.utcnow().isoformat(timespec="seconds")
+
+
+def _format_stack_from_frame(frame: FrameType, max_frames: int) -> str:
+    # Limit frames from the BOTTOM (most recent last) to keep prints short.
+    limit: int | None = max_frames if max_frames > 0 else None
+    return "".join(traceback.format_stack(frame, limit=limit))
+
+
+class AsyncioStallMonitor:
+    """
+    Hybrid stall monitor.
+
+    Detection path:
+      - In-loop heartbeat updates a shared timestamp.
+      - Background thread checks wall time vs. last heartbeat; if exceeds threshold,
+        we enter "stall" state, sample loop thread stack, and log.
+      - While stalled, we periodically re-sample (min_report_interval).
+      - When the loop catches up (heartbeat moves), we close the incident and
+        correlate with asyncio tasks inside the loop thread.
+    """
+
+    def __init__(
+        self,
+        loop: asyncio.AbstractEventLoop | None = None,
+        *,
+        threshold: float = 0.2,  # seconds without heartbeat => stall
+        heartbeat_interval: float = 0.1,  # loop posts heartbeat this often
+        min_report_interval: float = 0.5,  # during a long stall, log at most this often
+        max_frames: int = 25,  # limit stack frames captured
+        history_size: int = 1,  # TODO: increase and make history more useful in the future
+    ) -> None:
+        self.loop = loop or asyncio.get_event_loop()
+        self.threshold = float(threshold)
+        self.heartbeat_interval = float(heartbeat_interval)
+        self.min_report_interval = float(min_report_interval)
+        self.max_frames = int(max_frames)
+
+        self._hb_perf: float = perf_counter()
+        self._running = False
+        self._thread: threading.Thread | None = None
+        self._loop_thread_ident: int | None = None
+
+        self._incident: StallIncident | None = None
+        self._last_report_perf: float = 0.0
+
+        self._lock = threading.Lock()
+        self.history: deque[StallIncident] = deque(maxlen=history_size)
+
+    def start(self) -> None:
+        if self._running:
+            return
+        self._running = True
+
+        def _capture_thread_ident():
+            self._loop_thread_ident = threading.get_ident()
+
+        self.loop.call_soon(_capture_thread_ident)
+
+        # Kick off the heartbeat loop
+        self.loop.call_soon(self._heartbeat)
+
+        # Start watchdog thread
+        self._thread = threading.Thread(target=self._watchdog, name="asyncio-stall-watchdog", daemon=True)
+        self._thread.start()
+        log.info(
+            "AsyncioStallMonitor started: threshold=%.3fs heartbeat=%.3fs",
+            self.threshold,
+            self.heartbeat_interval,
+        )
+
+    def stop(self) -> None:
+        self._running = False
+        log.info("AsyncioStallMonitor stopping...")
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc, tb):
+        self.stop()
```

Comment:
do we need to do anything to ensure that this will terminate 

i know that Andrew Godwin originally did some special handling (see `_exit_gracefully` in trigger_job_runner)

this is sincere question  (rather than suggestion by question) -- maybe we don't need to do anything but just merits a thought

Reply:
I have improved the stop method